### PR TITLE
FileDao code changes implemented

### PR DESCRIPTION
**Pull request by tirthakharel**
This commit includes a new `File` package that has the following:

1.  `FileController` class for managing file related requests
2.  `FileType` enum for different types of files stored on the database
3.  `File` object type that details what the file looks like in the database
4.  Services for uploading, downloading, and deleting files (as well as various PDF specific services ported from the previous PDF package)

All testing of the services was completed manually with postman and the staging DB, when the unit and integration tests are completed will add on to this PR

The actual FileDao itself was made to be similar to the PDFDao we had previously - which was not used much since we relied heavily on the GridFSBucket functions for getting files - but can make more use of it if needed

### CODE REVIEW

File: src/main/Database/File/FileDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  @Override
+  public void save(File file) {}
+
+  @Override
+  public Optional<File> get(ObjectId id) {
+    return Optional.empty();
+  }
+
+  @Override
+  public Optional<File> get(String username) {
```
Comment (crchong1) [inline src/main/Database/File/FileDaoImpl.java:55]: wouldn't there be multiple files associated with a given username? maybe we should return a set<file>

File: src/main/File/Services/UploadFileService.java [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+  String filename;
+  String title;
+  String fileContentType;
+  String fileIdStr;
+  boolean annotated;
+  boolean toSign;
+  InputStream fileStream;
+  InputStream signatureFileStream;
+  FileType fileType;
+  MongoDatabase db;
```
Comment (crchong1) [inline src/main/File/Services/UploadFileService.java:58]: Why do we still have the MongoDatabase in this serrvice? Why can't we just move the implementation of saving the file to the filedao.save() function here? We should only include the fileDao in these services and let the dao handle the logic

File: src/main/File/Services/UploadFileService.java [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+      try {
+        return mongodbUploadMisc();
+      } catch (GeneralSecurityException | IOException e) {
+        return FileMessage.SERVER_ERROR;
+      }
+    }
+  }
+
+  public Message mongodbUploadPDF() throws GeneralSecurityException, IOException {
+    GridFSBucket gridBucket = GridFSBuckets.create(db, "files");
```
Comment (crchong1) [inline src/main/File/Services/UploadFileService.java:136]: you should rely on the filedao here

File: src/main/Database/File/FileDao.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import com.google.api.client.util.DateTime;
+import org.bson.types.ObjectId;
+
+import java.io.InputStream;
+import java.util.Optional;
+
+public interface FileDao extends Dao<File> {
+  Optional<File> get(ObjectId id);
+
+  Optional<File> get(String username);
```
Comment (crchong1) [inline src/main/Database/File/FileDao.java:15]: i think a getAll(String username) function would replace this

Comment (tirthakharel) [reply src/main/Database/File/FileDao.java]: Replaced in latest commit

File: src/main/File/File.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    this.username = username;
+    this.uploadedAt = uploadedAt;
+    this.fileStream = fileStream;
+    this.fileType = fileType;
+    this.filename = fileName;
+  }
+
+  public File(String username, Date uploadedAt, InputStream fileStream, FileType fileType) {
+    this.id = new ObjectId();
+    this.username = username;
```
Comment (crchong1) [inline src/main/File/File.java:91]: Just curious why this constructor exists if there is no file id?

File: src/main/File/File.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    this.uploadedAt = uploadedAt;
+    this.fileStream = fileStream;
+    this.fileType = fileType;
+  }
+
+  public File(String username, Date uploadedAt, ObjectId fileId, FileType fileType) {
+    this.id = new ObjectId();
```
Comment (crchong1) [inline src/main/File/File.java:98]: is this constructor necessary? I would prefer maybe just one or two constructors and we can just pass in Optional<> fields if we think those are not as applicable to every File

File: src/main/File/FileController.java [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+    try {
+      this.encryptionController = new EncryptionController(db);
+    } catch (Exception e) {
+      System.out.println("Error in encryption controller!");
+      e.printStackTrace();
+    }
+  }
+
+  public User userCheck(String req) {
+    log.info("Starting check for user...");
```
Comment (crchong1) [inline src/main/File/FileController.java:46]: lets use the UserDao here instead of calling mongodb?

File: src/main/File/FileController.java [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+
+import java.io.InputStream;
+import java.util.Objects;
+
+import static User.UserController.mergeJSON;
+import static com.mongodb.client.model.Filters.eq;
+
+@Slf4j
+public class FileController {
+  private MongoDatabase db;
```
Comment (crchong1) [inline src/main/File/FileController.java:28]: We can probably remove the MongoDatabase variable and rely on the daos

File: src/main/File/FileController.java [hunk 9]
```diff
@@ -0,0 +1,478 @@
...
+              boolean toSign = false;
+              if (ctx.formParam("annotated") != null) {
+                annotated = Boolean.parseBoolean(ctx.formParam("annotated"));
+              }
+              if (ctx.formParam("toSign") != null) {
+                toSign = Boolean.parseBoolean(ctx.formParam("toSign"));
+              }
+              String fileId = null;
+              UploadedFile signature = null;
+              if (fileType.isPDF()) {
```
Comment (crchong1) [inline src/main/File/FileController.java:132]: can we use a switch statement here? Just so we can easily tell if we need to add another file type

File: src/main/File/FileController.java [hunk 10]
```diff
@@ -0,0 +1,478 @@
...
+                  fileId = Objects.requireNonNull(ctx.formParam("fileID"));
+                }
+                UploadFileService uploadService =
+                    new UploadFileService(
+                        db,
+                        fileDao,
+                        username,
+                        organizationName,
+                        privilegeLevel,
+                        fileType,
```
Comment (crchong1) [inline src/main/File/FileController.java:158]: To me this constructor seems a bit too long. Can't we just pass the File object here and the UploadFileService can just get the fields out of the File?

File: src/main/File/FileController.java [hunk 11]
```diff
@@ -0,0 +1,478 @@
...
+        UserType userType;
+        JSONObject req = new JSONObject(ctx.body());
+        User check = userCheck(ctx.body());
+        if (check == null && req.has("targetUser")) {
+          log.info("Target User not Found");
+          ctx.result(UserMessage.USER_NOT_FOUND.toJSON().toString());
+        } else {
+          boolean orgFlag;
+          if (check != null && req.has("targetUser")) {
+            log.info("Target user found");
```
Comment (crchong1) [inline src/main/File/FileController.java:241]: Shouldn't this logic exist in another service somewhere?

File: src/main/File/Services/DeleteFileService.java [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+  MongoDatabase db;
+  private FileDao fileDao;
+  private String username;
+  private String orgName;
+  private UserType userType;
+  private FileType fileType;
+  private String fileId;
+
+  public DeleteFileService(
+      MongoDatabase db,
```
Comment (crchong1) [inline src/main/File/Services/DeleteFileService.java:26]: as stated before, lets remove the MongoDatabase db field and just rely on the DAO

File: src/main/File/Services/DownloadFileService.java [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.GeneralSecurityException;
+import java.util.Objects;
+import java.util.Optional;
+
+@Slf4j
+public class DownloadFileService implements Service {
+  MongoDatabase db;
```
Comment (crchong1) [inline src/main/File/Services/DownloadFileService.java:23]: as stated before, lets remove the MongoDatabase db field and just rely on the DAO

File: src/main/File/Services/FillPDFFileService.java [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+public class FillPDFFileService implements Service {
```
Comment (crchong1) [inline src/main/File/Services/FillPDFFileService.java:20]: just wondering does this new service use the FileDao at all?

File: src/main/File/Services/FillPDFFileService.java [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+  UserType privilegeLevel;
+  InputStream fileStream;
+  JSONObject formAnswers;
+  InputStream completedForm;
+  MongoDatabase db;
```
Comment (crchong1) [inline src/main/File/Services/FillPDFFileService.java:25]: remove this field

File: src/main/File/Services/GetFilesInformationService.java [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+import java.util.Objects;
+
+import static com.mongodb.client.model.Filters.and;
+import static com.mongodb.client.model.Filters.eq;
+
+public class GetFilesInformationService implements Service {
+  MongoDatabase db;
```
Comment (crchong1) [inline src/main/File/Services/GetFilesInformationService.java:21]: remove

File: src/main/File/Services/UploadFileService.java (hunk 2 above)
Comment (crchong1) [reply src/main/File/Services/UploadFileService.java:47]: its still there?

File: src/main/User/UserController.java [hunk 17]
```diff
@@ -293,6 +293,24 @@ public static JSONObject mergeJSON(
         UploadedFile file = ctx.uploadedFile("file");
         log.info(username + " is attempting to upload a profile picture");
         UploadPfpService serv = new UploadPfpService(db, username, file, fileName);
+        // TODO: enable UploadFileService:
+        //        UploadFileService serv =
+        //            new UploadFileService(
+        //                db,
+        //                username,
+        //                null,
```
Comment (crchong1) [inline src/main/User/UserController.java:301]: I would like to move away from null constructor arguments- could we maybe just use Optional?

File: src/test/DatabaseTest/File/FileDaoImplUnitTests.java [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+
+  @After
+  public void reset() {
+    fileDao.clear();
+  }
+
+  @Test
+  public void save() {
+    String testUsername = "username1";
+    File file = EntityFactory.createFile().withUsername(testUsername).build();
```
Comment (crchong1) [inline src/test/DatabaseTest/File/FileDaoImplUnitTests.java:34]: can we assert on every possible field in the FileDao? Just to make sure

File: src/test/FileTest/FileControllerIntegrationTests.java [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+          + "src"
+          + File.separator
+          + "test"
+          + File.separator
+          + "resources";
+
+  @BeforeClass
+  public static void setUp() {
+    TestUtils.startServer();
+    TestUtils.setUpTestDB();
```
Comment (crchong1) [inline src/test/FileTest/FileControllerIntegrationTests.java:40]: i dont think you need to set up the test db here btw- lets try to use the entity factory which allows us to start from a blank slate

**Review by crchong1 — APPROVED** [review]
once the comments are addressed, feel free to merge

File: src/main/File/File.java (hunk 5 above)
Comment (tirthakharel) [reply src/main/File/File.java:91]: removed

**Pull request merged**